exonum/src/storage/sparse_list_index.rs

```diff
@@ -0,0 +1,735 @@
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use exonum::storage::{MemoryDB, Database, SparseListIndex};
+    ///
+    /// let db = MemoryDB::new();
+    /// let mut fork = db.fork();
+    /// let mut index = SparseListIndex::new(vec![1, 2, 3], &mut fork);
+    /// assert_eq!(None, index.last());
+    ///
+    /// index.push(42);
+    /// assert_eq!(Some(42), index.last());
+    /// ```
+    pub fn last(&self) -> Option<V> {
+        match self.len() {
+            0 => None,
+            l => self.get(l - 1),
+        }
+    }
+
+    /// Returns `true` if the list contains no elements.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use exonum::storage::{MemoryDB, Database, SparseListIndex};
+    ///
+    /// let db = MemoryDB::new();
+    /// let mut fork = db.fork();
+    /// let mut index = SparseListIndex::new(vec![1, 2, 3], &mut fork);
+    /// assert!(index.is_empty());
+    ///
+    /// index.push(42);
+    /// assert!(!index.is_empty());
+    /// ```
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+
+    /// Returns the total number of elements (incudes null elements) in the list.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use exonum::storage::{MemoryDB, Database, SparseListIndex};
+    ///
+    /// let db = MemoryDB::new();
+    /// let mut fork = db.fork();
+    /// let mut index = SparseListIndex::new(vec![1, 2, 3], &mut fork);
+    /// assert_eq!(0, index.len());
+    ///
+    /// index.push(10);
+    /// index.push(12);
+    /// assert_eq!(2, index.len());
+    ///
+    /// index.remove(0);
+    ///
+    /// index.push(100);
+    /// assert_eq!(3, index.len());
+    /// ```
+    pub fn len(&self) -> u64 {
+        self.size().total_length
+    }
+
+    /// Returns the total number of non-null elements in the list.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use exonum::storage::{MemoryDB, Database, SparseListIndex};
+    ///
+    /// let db = MemoryDB::new();
+    /// let mut fork = db.fork();
+    /// let mut index = SparseListIndex::new(vec![1, 2, 3], &mut fork);
+    /// assert_eq!(0, index.count());
+    ///
+    /// index.push(10);
+    /// assert_eq!(1, index.count());
+    ///
+    /// index.remove(0);
+    ///
+    /// index.push(100);
+    /// assert_eq!(1, index.count());
+    /// ```
+    pub fn count(&self) -> u64 {
+        self.size().items_count
+    }
+
+    /// Returns an iterator over the list. The iterator element type is V.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use exonum::storage::{MemoryDB, Database, SparseListIndex};
+    ///
+    /// let db = MemoryDB::new();
+    /// let mut fork = db.fork();
+    /// let mut index = SparseListIndex::new(vec![1, 2, 3], &mut fork);
+    ///
+    /// index.extend([1, 2, 3, 4, 5].iter().cloned());
+    ///
+    /// for val in index.iter() {
+    ///     println!("{}", val);
+    /// }
+    /// ```
+    pub fn iter(&self) -> SparseListIndexIter<V> {
```

Comment:
Since sparse list is less trivial than a list, I think that its iterators should be more like the iterators of map:

indices() -> iterator by indeces (like keys in maps)
values() -> iterator by values
iter() -> iterator by pairs (index, value)